spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
-        Get the current color scheme.
+        # Actions to add to the Options menu
+        self._plugin_actions = None
 
-        Returns
-        -------
-        dict
-            Dictionary with properties and colors of the color scheme
-            used in the Editor.
+        # Attribute to keep track if the plugin is undocked in a
+        # separate window
+        self._undocked_window = None
 
-        Notes
-        -----
-        This is useful to set the color scheme of all instances of
-        CodeEditor used by the plugin.
-        """
-        return super(BasePluginWidget, self)._get_color_scheme()
+        self._ismaximized = False
+        self._default_margins = None
+        self._isvisible = False
 
-    def switch_to_plugin(self):
-        """
-        Switch to this plugin.
+        self.shortcut = None
 
-        Notes
-        -----
-        This operation unmaximizes the current plugin (if any), raises
-        this plugin to view (if it's hidden) and gives it focus (if
-        possible).
-        """
-        super(BasePluginWidget, self)._switch_to_plugin()
+        # Options buttons
+        self.options_button = create_toolbutton(self, text=_('Options'),
+                                                icon=ima.icon('tooloptions'))
+        self.options_button.setPopupMode(QToolButton.InstantPopup)
 
+        # Don't show menu arrow and remove padding
+        if is_dark_interface():
+            self.options_button.setStyleSheet(
+                ("QToolButton::menu-indicator{image: none;}\n"
+                 "QToolButton{padding: 3px;}"))
+        else:
+            self.options_button.setStyleSheet(
+                "QToolButton::menu-indicator{image: none;}")
+
+        # Options menu
+        self._options_menu = QMenu(self)
+
+        # We decided to create our own toggle action instead of using
+        # the one that comes with dockwidget because it's not possible
+        # to raise and focus the plugin with it.
+        self._toggle_view_action = None
+
+        # Default actions for Options menu
+        self._dock_action = create_action(
+            self,
+            _("Dock"),
+            icon=ima.icon('dock'),
+            tip=_("Dock the pane"),
+            triggered=self._close_window,
+        )
+
+        self._undock_action = create_action(
+            self,
+            _("Undock"),
+            icon=ima.icon('undock'),
+            tip=_("Undock the pane"),
+            triggered=self._create_window,
+        )
+
+        self._close_plugin_action = create_action(
+            self,
+            _("Close"),
+            icon=ima.icon('close_pane'),
+            tip=_("Close the pane"),
+            triggered=self._plugin_closed,
+        )
+
+    # INFO: This is called from mainwindow.py
+    # This name does not convey that this can also (and should) be called
+    # from a non dockable widget. Non dockable widgets should be able to
+    # define a on_first_registration call
```

Comment:
Why? What could they need to do on first registration?

Reply:
A kite plugin would want t to show a popup and explain the user something, or open a `kite_tutorial.py`. This is just an example of something that NOW does need this.